File: bin/install_packages.rb
# frozen_string_literal: true

require_relative './distro'

def ensure_sudo
  sudo_path = '/usr/bin/sudo'
  return if File.exist?(sudo_path)

  p 'error expected ${sudoPath} to exist. unsupported system'
  exit 1
end

def install_macos_packages(config_dir)
  system('/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"')
  packages_file = File.readlines("#{config_dir}/packages/macos.txt")
  packages_file.each do |package|
    check = system("brew ls --versions #{package}")
    system("brew #{check ? 'upgrade' : 'install'} #{package}")
  end
end

def install_debian_packages(config_dir)
  system('sudo apt-get update')
  system('sudo apt-get install -y git ruby make locales')
  packages_file = File.readlines("#{config_dir}/packages/debian.txt")
  packages_file.each do |package|
    system("sudo apt-get install -y #{package}")
  end
  system('sudo apt-get autoremove')
  system('sudo apt-get autoclean')
  system('sudo locale-gen en_US.UTF-8')
end

def install_fedora_packages(config_dir)
  system('dnf install -y git ruby make ')
  system('sudo dnf group install "C Development Tools and Libraries"')
  packages_file = File.readlines("#{config_dir}/packages/fedora.txt")
  packages_file.each do |package|
    system("sudo dnf install -y #{package}")
  end
end

def install_packages
  ensure_sudo
  config_dir = "#{File.dirname(__FILE__)}/.."

  return install_macos_packages(config_dir) if darwin?
  return install_debian_packages(config_dir) if debian?
  return install_fedora_packages(config_dir) if fedora?

  puts 'fatal: unsupported system'
  exit 1
end

install_packages if __FILE__ == $PROGRAM_NAME

File: bin/install_packages.ts
#!/usr/bin/env deno run --allow-read --allow-write --allow-env --allow-net --allow-run
import { existsSync, readFileStrSync } from "https://deno.land/std/fs/mod.ts";
import { isDarwin, isDebian, isFedora } from "./distro.ts";

export async function installPackages() {
  const sudoPath = "/usr/bin/sudo";

  if (!existsSync(sudoPath)) {
    console.log(`error expected ${sudoPath} to exist. unsupported system`);
    Deno.exit(1);
  }

  let dirname = import.meta.url
    .substring(0, import.meta.url.lastIndexOf("/"))
    .replace("file:/", "");
  const configDir = `${dirname}/..`;

  const darwin = await isDarwin();
  if (darwin) {
    const packagesFile = readFileStrSync(`${configDir}/packages/macOS`);
    const packages = packagesFile.trim().split("\n");

    for (const pkg of packages) {
      const check = await Deno.run({ cmd: ["brew", "ls", "--versions", pkg] }).status();
      await Deno.run({ cmd: ["brew", check.success ? "upgrade" : "install", pkg] }).status();
    }
  } else if (isDebian()) {
    const pm = isDebian() ? "apt-get" : "dnf";
    await Deno.run({ cmd: ["sudo", pm, "update", "-y"] }).status();

    const packagesFilePath = isDebian() ? "packages/Debian" : "packages/Fedora";
    const packages = readFileStrSync(`${configDir}/${packagesFilePath}`)
      .split("\n");

    await Deno.run({
      cmd: ["sudo", pm, "install", "-y"].concat(packages)
    }).status();
    await Deno.run({ cmd: ["sudo", pm, "autoremove"] }).status();
    await Deno.run({ cmd: ["sudo", pm, "autoclean"] }).status();

    await Deno.run({ cmd: ["sudo", "locale-gen", '"en_US.UTF-8"'] }).status();
  }
}

if (import.meta.main) {
  (async () => {
    await installPackages();
  })();
}
